menhir-recover/recovery_new.ml

```diff
@@ -0,0 +1,196 @@
+        let item = st, prod, pos in
+        let cost, _ = S.solve (S.Tail(st, prod, pos)) in
+        (Production.lhs prod, {cost = cost; items = [item]}) :: acc
+      ) [] (Lr0.items (Lr1.lr0 st))
+    )
+
+  let step child parent =
+    let finish = List.filter_map (fun (prod, pos) ->
+        if Array.length (Production.rhs prod) <= pos then
+          None
+        else 
+          match (Production.rhs prod).(pos) with
+            | (N nt, _, _) -> 
+              let next = List.assoc (N nt) (Lr1.transitions parent) in
+              let cost, _ = S.solve (S.Tail(next, prod, pos + 1)) in
+              Some (nt, {cost = cost; items = [next, prod, pos + 1]})
+            | _ -> None
+      ) @@ Lr0.items (Lr1.lr0 parent) in
+    
+    let state = State.empty () in
+
+    List.iter (State.add_elem state) @@ synthesize child;
+
+    let rec aux acc =
+      if State.is_empty state then
+        acc
+      else 
+        let (nt, tr) = State.get_min state in
+        match List.find_opt (fun (nt', _) -> nt = nt') finish with
+          | Some (_, tr') -> 
+            aux (Trace.cat tr' tr :: acc)
+          | _ ->
+            (* TODO workaround for children with different positions in items*)
+            match List.assoc_opt (N nt) (Lr1.transitions parent) with 
+              | None -> aux acc 
+              | Some st' ->
+                synthesize st'
+                |> List.map (fun (nt, tr') -> nt, Trace.cat tr' tr)
+                |> List.iter @@ State.add_elem state;
+                aux acc
+    in
+    (List.fold_left Trace.min {cost = Cost.infinite; items = []} @@ aux []).items
+
+  let pred = Lr1.tabulate 
+    (fun st -> 
+      Lr1.fold (fun st' acc -> 
+        if List.fold_left (fun flag (_, st'') -> st = st'' || flag) false (Lr1.transitions st') then
+          st' :: acc
+        else
+          acc
+      ) []
+    )
+
+  let parents st = 
+
+    let preds states = 
+      List.map pred states
+      |> List.flatten
+      |> List.sort_uniq compare in
+
+    let rec aux states = function
+      | 0 -> []
+      | 1 -> preds states
+      | n -> aux (preds states) (n - 1) in
+
+    aux [st] (depth st)
+
+  let has_different_pos st =
+    let items = Lr0.items (Lr1.lr0 st) in
+    List.length items > 0 &&
+    List.length @@ group_assoc @@ List.map (fun (a, b) -> (b, a)) items > 1
+
+  let cheapest_item st =
+    let items = List.map (fun (prod, pos) ->
+      let item = st, prod, pos in
+      let cost, _ = S.solve (S.Tail(st, prod, pos)) in
+      (cost, item)
+    ) (Lr0.items (Lr1.lr0 st)) in
+    let (_, cheapest_item) = List.fold_left
+      (fun ((costx, (_, _, posx)) as x) ((costy, (_, _, posy)) as y) ->
+        if (costx, - posx) < (costy, - posy) then x else y (* Pos comparation is needed for state 430 in PascaLIGO*)
+    ) (List.hd items) items in
+    cheapest_item
+  
+  let recover st : recovery =
+    { 
+      prefix = 100500; (*Doesn't matter*)
+      cases = 
+      let pars = parents st in
```

Comment:
It would be better to move all logic out from the record definition and put it above it.